src/maindomaintest.ml

```diff
@@ -66,46 +66,72 @@ let domains: (module Lattice.S) list = [
 
 let nonAssocDomains: (module Lattice.S) list = []
 
-let intDomains: (module IntDomain.IkindUnawareS) list = [
-  (module IntDomain.Flattened);
+let intDomains: (module IntDomainProperties.S) list = [
+  (module IntDomain.Interval);
+  (* (module IntDomain.Flattened); *)
   (* (module IntDomain.Interval32); *)
   (* (module IntDomain.Booleans); *)
   (* (module IntDomain.CircInterval); *)
   (* (module IntDomain.Enums); *)
   (* (module IntDomain.IntDomTuple); *)
 ]
 
-let nonAssocIntDomains: (module IntDomain.IkindUnawareS) list = [
-  (* (module IntDomain.DefExc) *)
+let nonAssocIntDomains: (module IntDomainProperties.S) list = [
+  (module IntDomain.DefExc);
+]
+
+(* TODO: make arbitrary ikind part of domain test for better efficiency *)
+let ikinds: Cil.ikind list = [
+  (* TODO: enable more, some seem to break things *)
+  IChar;
+  ISChar;
+  (* IUChar; *)
+  (* IBool; *)
+  IInt;
+  (* IUInt; *)
+  IShort;
+  (* IUShort; *)
+  (* ILong; *)
+  (* IULong; *)
+  (* ILongLong; *)
+  (* IULongLong; *)
 ]
 
 let testsuite =
-  List.map (fun d ->
+  domains
+  |> List.concat_map (fun d ->
       let module D = (val d: Lattice.S) in
       let module DP = DomainProperties.All (D) in
-      DP.tests)
-    domains
-  |> List.flatten
+      DP.tests
+    )
 let nonAssocTestsuite =
-  List.map (fun d ->
+  nonAssocDomains
+  |> List.concat_map (fun d ->
       let module D = (val d: Lattice.S) in
       let module DP = DomainProperties.AllNonAssoc (D) in
-      DP.tests)
-    nonAssocDomains
-  |> List.flatten
+      DP.tests
+    )
+
+let old_intdomains intDomains =
+  BatList.cartesian_product intDomains ikinds
+  |> List.map (fun (d, ik) ->
+      let module D = (val d: IntDomainProperties.S) in
+      let module Ikind = struct let ikind () = ik end in
+      (module IntDomainProperties.WithIkind (D) (Ikind): IntDomainProperties.OldS)
+    )
 let intTestsuite =
-  List.map (fun d ->
-      let module D = (val d: IntDomain.IkindUnawareS) in
+  old_intdomains intDomains
+  |> List.concat_map (fun d ->
```

Comment:
Which version of OCaml are we aiming at now? This is `Ocaml >= 4.10.0`.
In the CI we are pretending to check (but failing)  support for `4.07.1` too.
Is support for `4.07.1` something we want to abandon?

Reply:
I guess this file could also `open Prelude`, which is our own and opens `Batteries`, which gives `List = BatList`, which should have `concat_map` regardless.

Reply:
> In the CI we are pretending to check (but failing) support for `4.07.1` too.

Also, what do you mean by that?

Reply:
No, `BatList` from the latest version (`batteries.3.3.0`) also does not have `concat_map` for `OCaml < 4.10.0`.
https://ocaml-batteries-team.github.io/batteries-included/hdoc2/BatList.html

Reply:
Damn, that's sure some weird choice by Batteries to limit unnecessarily availability also by OCaml version. I guess I can revert to `map` and `concat` then.

Reply:
> Also, what do you mean by that?

As evidenced by this PR, we do not really compile with `4.07.1`. We specify it in the Github actions

https://github.com/goblint/analyzer/blob/16497fa06d2059f2c3bde130da0bbb89b5a4d6c3/.github/workflows/tests.yml#L17-L19

but then our CI script uses the `pinned` file, installing the pinned version of OCaml.

Reply:
If we want to still support 4.07 we should somehow fix the CI then, otherwise it's super misleading and pointless. I don't know how that would work with locking though.

So it seems easier to just ditch all the OCaml version stuff from the CI and just test the locked version. I don't have a reason for still supporting 4.07, but if someone does, I guess the safe thing then would be to have 4.07 locked instead. There's no use to locking 4.11 if none of the new features can be used.

Or I guess a mix of the two is also possible with the following CI configurations:
* Install locked dependencies (including OCaml)
* Install unlocked dependencies and OCaml 4.07
* Install unlocked dependencies and OCaml 4.11

Reply:
Yes, I agree we should fix this.

**General considerations**
Ideally I was hoping to go for a setup like I have for CIL, where we just do everything via `opam` and `dune`.

https://github.com/goblint/cil/blob/60e9c5f77a7267591622b94c9be73555cc2881f1/.github/workflows/tests.yml

This is by far the cleanest and would also give us an indication if some package in an updated version is incompatible with Goblint such that we can put a constraint on the maximum version number of some package. Additional advantage is that testing multiple versions of OCaml is a breeze.

Downside is that we would not have reproduceable builds as the version of dependencies might change between runs.

**Now, as for support of 4.07.1 is concerned:** 
@jerhard needed this at some point for `earlybird`, but I guess `earylbird` now supports later versions? So maybe we can drop support for 4.07.1 now? However, I would still want to keep compatibility with some older versions (e.g. 4.09.1 now), because in my experience it sometimes takes time for e.g. the VSCode plugin (and the language server) to be adapted to a new version of merlin and it is annoying when such things break but you can't go back because Goblint uses new features.

Reply:
I'll just replace these `List.concat_map`s for now, so we can merge this while still being 4.07 compatible.

We should probably move this CI fixing to a new issue. Cleaning up the CI procedures is something worth doing regardless of what minimum version we want to keep for now, because it's still `travis-ci.sh`...

Reply:
>@jerhard needed this at some point for earlybird, but I guess earylbird now supports later versions? So maybe we can drop support for 4.07.1 now?

I used OCaml 4.07.1 for some time, to be able to use it together with `earlybird`, yes. Currently I use 4.11.1, so I don't rely on the compatibility with 4.07.1 anymore.
The last time I tried, I did not get `earlybird` to work with Goblint and the 4.11.1 compiler, however.

Reply:
The issue is now tracked as #184.